### PR TITLE
Feature/add cinema hall request specs

### PR DESCRIPTION
**Pull request by atytko**
- Added factories for user roles, cinema halls, and users.
- Created cinema hall controller.
- Added request specs for cinema hall controller
- Set up routes for cinema hall.
- Added tokens in headers in auth_helper.rb (since devise and devise_token_auth have different method of authentification).
(thanks Wojtek!)

### CODE REVIEW

File: app/controllers/cinema_halls_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+class CinemaHallsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :find_cinema_hall, only: %i[show update destroy]
```
Comment (luktuc) [inline app/controllers/cinema_halls_controller.rb:5]: This is not wrong. But it's really hard to follow. So when I look for a show method I need to look above for a before_action and then to private to know what's going on. Sometimes DRY is not a best choice. https://blog.ragnarson.com/2015/09/30/problems-with-typical-rails-controllers-and-before-actions.html

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -33,6 +34,8 @@
   exit 1
 end
 RSpec.configure do |config|
+  config.include AuthHelpers::Includables, type: :request
+  config.extend AuthHelpers::Extensions, type: :request
```
Comment (jandudulski) [inline spec/rails_helper.rb:38]: Common practice is to move such extensions directly to the support file. Like:

```
# support/auth_helpers.rb
module AuthHelpers
  # ...
end

RSpec.configure do |config|
  config.include ...
end
```

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 # This file is copied to spec/ when you run 'rails generate rspec:install'
 require 'spec_helper'
+require "support/auth_helper"
```
Comment (jandudulski) [inline spec/rails_helper.rb:5]: When you uncomment line 26 `# Dir[Rails.root.join('spec', 'support', '**', '*.rb')].sort.each { |f| require f }` it would automatically require all files from `support` directory

File: spec/requests/cinema_halls_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+require 'rails_helper'
+require 'shared_examples_for_cinema_halls'
+
+RSpec.describe CinemaHallsController, type: :request do
+  let(:role) { create(:user_role) }
+  let(:user) { create(:user, user_role: role) }
+
+  describe '#index' do
+    subject { get '/cinema_halls' }
```
Comment (jandudulski) [inline spec/requests/cinema_halls_controller_spec.rb:11]: Subjects might have a descriptive name, like: `subject(:get_cinema_halls) { .. }` and later on in the test, you can use that name instead of `subject` which might be quite cryptic in larger files.

File: spec/shared_examples_for_cinema_halls.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+RSpec.shared_examples 'unauthorized' do |_parameter|
+  let(:role) { create(:user_role) }
+  let(:user) { create(:user, user_role: role) }
```
Comment (jandudulski) [inline spec/shared_examples_for_cinema_halls.rb:5]: What's the purpose of adding this lets? You are not using them as far as I see.

File: spec/support/auth_helper.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      let(:auth_helpers_auth_token) do
+        public_send(user).create_new_auth_token
+      end
+    end
+  end
+
+  module Includables
+    HTTP_HELPERS_TO_OVERRIDE =
+      %i[get post patch put delete].freeze
+    # Override helpers for Rails 5.0
```
Comment (jandudulski) [inline spec/support/auth_helper.rb:15]: AFAIK you're using Rails 6. Does this still apply?

File: spec/rails_helper.rb (hunk 2 above)
Comment (atytko) [reply spec/rails_helper.rb]: thanks, I have refactored that part.

File: spec/requests/cinema_halls_controller_spec.rb (hunk 4 above)
Comment (atytko) [reply spec/requests/cinema_halls_controller_spec.rb]: Thanks, I have refactored subjects 👍

File: spec/shared_examples_for_cinema_halls.rb (hunk 5 above)
Comment (atytko) [reply spec/shared_examples_for_cinema_halls.rb]: Classic opy and paste mistake :D I have removed it, thank you!

File: app/controllers/cinema_halls_controller.rb (hunk 1 above)
Comment (atytko) [reply app/controllers/cinema_halls_controller.rb:5]: Is the below code something you have in mind? 🤔 

`# frozen_string_literal: true

class CinemaHallsController < ApplicationController
  before_action :authenticate_user!

  def index
    cinema_halls = CinemaHall.all

    render json: cinema_halls
  end

  def show
    find_cinema_hall
    render json: @cinema_hall
  end

  def create
    cinema_hall = CinemaHall.new(cinema_hall_params)
    if cinema_hall.save
      render json: cinema_hall, status: :created, location: cinema_hall
    else
      render json: cinema_hall.errors, status: :unprocessable_entity
    end
  end

  def update
    find_cinema_hall
    if @cinema_hall.update(cinema_hall_params)
      render json: @cinema_hall
    else
      render json: @cinema_hall.errors, status: :unprocessable_entity
    end
  end

  def destroy
    find_cinema_hall
    @cinema_hall.destroy
    head :no_content
  end

  private

  def find_cinema_hall
    @cinema_hall = CinemaHall.find(params[:id])
  end

  def cinema_hall_params
    params.require(:cinema_hall).permit(:name, :row_number, :row_total_seats)
  end
end`

File: spec/support/auth_helper.rb (hunk 6 above)
Comment (atytko) [reply spec/support/auth_helper.rb:15]: I think so 👀  At least it helped to fix the error preventing the tests from working properly.

Comment (wmaciejak) [reply spec/support/auth_helper.rb]: This is my solution which I've partly found on the internet cause we had to solve problems with authentication in specs somehow. Generally, it still applies to Rails 6.0 👍

File: config/environments/test.rb [hunk 7]
```diff
@@ -45,6 +45,8 @@
   # ActionMailer::Base.deliveries array.
   config.action_mailer.delivery_method = :test
 
+  config.action_mailer.default_url_options = {:host => "localhost:3000"}
```
Comment (jandudulski) [inline config/environments/test.rb:48]: Most people prefer this syntax: `{host: "localhost:3000"}` ;)

Comment (atytko) [reply config/environments/test.rb]: thanks, corrected. 👍

Comment (jandudulski) [reply config/environments/test.rb]: Actually, it's nothing wrong with the previous version - both are correct and doing the same. It's just that it's good to be consistent and stick to one version :)

File: app/controllers/cinema_halls_controller.rb (hunk 1 above)
Comment (atytko) [reply app/controllers/cinema_halls_controller.rb]: refactored, thanks :)

File: app/controllers/cinema_halls_controller.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  before_action :authenticate_user!
+
+  def index
+    cinema_halls = CinemaHall.all
+
+    render json: cinema_halls
+  end
+
+  def show
+    CinemaHall.find(params[:id])
```
Comment (jandudulski) [inline app/controllers/cinema_halls_controller.rb:13]: You forgot to assign it to the variable, so your render with return nothing now.

Comment (atytko) [reply app/controllers/cinema_halls_controller.rb]: Thanks, corrected 👍

**Review by wmaciejak — APPROVED** [review]
Really cool usage of shared examples 👏
